usb_winusb.go

```diff
@@ -58,19 +59,20 @@ func winusbError(err error) error {
 	return nil
 }
 
-func usbopen(cid string) (err error) {
+func usbopen(cid string, timeout time.Duration, serial string) (err error) {
 	if device.ctx != nil {
 		log.WithField("Correlation-ID", cid).Debug("usb context already open")
 		return nil
 	}
 
+	ms := C.ulong(timeout.Milliseconds())
```

Comment:
On further reading / thinking... I'm not sure if the data size for PIPE_TRANSFER_TIMEOUT _must_ be 32-bits (sizeof(ULONG) on Windows), or if it can handle 64 bits. Some testing might be required.

If it does need 32 bits... then I guess you can convert on either side - Go or C. It's unclear if C.ulong() will throw any fuss when an integer overflow occurs. Judging by the lack of err returned, I'm guessing it's as if you do a simple "C-style" cast. (i.e. silent overflow)

So you might want to do it on the C / Windows side. There's the intsafe.h libraries - the function you'd want is [ULongLongToULong](https://docs.microsoft.com/en-us/windows/win32/api/intsafe/nf-intsafe-ulonglongtoulong) (ULONGLONG is guaranteed to be 64-bit on all platforms / architectures). Check to make sure the function succeeds, return error otherwise.

Reply:
Great point! I changed the type of timeout to uint32 because realistically the users will never need to specify a timeout bigger than that (which is already 1193 hours!).
Viper guarantees that the number fits in a uint32, otherwise it erros out, so before invoking `C.ulong` the `timeout` (a `int64`) is cast to `uint32` in go land, and if I understand correctly that should always fit in a `ULONG`.

Reply:
Realistically, this is probably a fine resolution. As you said, in practice - the timeouts are never going to get this large. I'm still slightly bothered by the _implicit_ cast from int64 -> uint32. Can we place an assert or a debug log message to catch the one-in-a-million chance someone changes the code in the future and causes an inadvertent overflow? Call me paranoid, but I've been saved by things like this in the past.

Reply:
Agree completely! Add assert to make sure the value fits in a uint32.